src/NLog/Internal/FileAppenders/BaseFileAppender.cs

```diff
@@ -158,6 +174,78 @@ protected void FileTouched(DateTime dateTime)
             this.LastWriteTime = dateTime;
         }
 
+#if !SILVERLIGHT
+        /// <summary>
+        /// Creates a mutually-exclusive lock for archiving files.
+        /// </summary>
+        /// <returns>A <see cref="Mutex"/> object which can be used for controlling the archiving of files.</returns>
+        protected virtual Mutex CreateArchiveMutex()
+        {
+            return new Mutex();
+        }
+
+        /// <summary>
+        /// Creates a mutex for archiving that is sharable by more than one process.
+        /// </summary>
+        /// <returns>A <see cref="Mutex"/> object which can be used for controlling the archiving of files.</returns>
+        protected Mutex CreateSharableArchiveMutex()
+        {
+            return CreateSharableMutex("FileArchiveLock");
+        }
+
+        /// <summary>
+        /// Creates a mutex that is sharable by more than one process.
+        /// </summary>
+        /// <param name="mutexNamePrefix">The prefix to use for the name of the mutex.</param>
+        /// <returns>A <see cref="Mutex"/> object which is sharable by multiple processes.</returns>
+        protected Mutex CreateSharableMutex(string mutexNamePrefix)
+        {
+            // Creates a mutex sharable by more than one process
+            var mutexSecurity = new MutexSecurity();
+            var everyoneSid = new SecurityIdentifier(WellKnownSidType.WorldSid, null);
+            mutexSecurity.AddAccessRule(new MutexAccessRule(everyoneSid, MutexRights.FullControl, AccessControlType.Allow));
+
+            // The constructor will either create new mutex or open
+            // an existing one, in a thread-safe manner
+            bool createdNew;
+            return new Mutex(false, GetMutexName(mutexNamePrefix), out createdNew, mutexSecurity);
+        }
+
+        private string GetMutexName(string mutexNamePrefix)
+        {
+            const string mutexNameFormatString = @"Global\NLog-File{0}-{1}";
+            const int maxMutexNameLength = 260;
+
+            string canonicalName = Path.GetFullPath(FileName).ToLowerInvariant();
+
+            // Mutex names must not contain a backslash, it's the namespace separator,
+            // but all other are OK
+            canonicalName = canonicalName.Replace('\\', '/');
+            string mutexName = string.Format(mutexNameFormatString, mutexNamePrefix, canonicalName);
+
+            // A mutex name must not exceed MAX_PATH (260) characters
+            if (mutexName.Length <= maxMutexNameLength)
+            {
+                return mutexName;
+            }
+
+            // The unusual case of the path being too long; let's hash the canonical name,
+            // so it can be safely shortened and still remain unique
+            string hash;
+            using (MD5 md5 = MD5.Create())
```

Comment:
Yes, this code came from `MutexMultiProcessFileAppender`. Adding a Guid won't work, because a different process will compute a different Guid for the same file. We need the same file to have the same mutex-name across processes.